### PR TITLE
Kk velero

### PR DESCRIPTION
**Pull request by katarzyna-koltun-mx**

### CODE REVIEW

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+linktitle: "Use Velero to Back Up Namespaces"
+url: /developerportal/deploy/private-cloud-velero/
+description: "Describes the process for using Velero to create and restore backups of your Mendix app namespaces in private cloud"
+weight: 20
+tags: ["Deploy", "Private Cloud", "Backup", "Velero", "Kubernetes"]
+---
+
+## 1 Introduction
+
+[Velero](https://velero.io/docs/) is a tool that you can use to back up your Kubernetes namespaces. It does not back up your database or S3 resources, so it cannot serve as your main backup method, but it can supplement the default [backup process](/developerportal/operate/backups/) for your Mendix app.
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: @Nidhi251289 Are there any particular use cases for when someone might want to use Velero? Or when it would not be appropriate?

Comment (Nidhi251289) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: velero is itself a use case.. when some one wants to back up their kubernetes resources

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    {{% alert color="info" %}}The process of creating a recovery cluster may vary depending on the platform that you use to host your private cloud. For more information, refer to the documentation supplied by your cloud provider.{{% /alert %}}
+
+## 3 Creating a Velero Backup
+
+To create a backup with Velero, follow these steps:
+
+1. Stop the Mendix Operator and Mendix agent by scaling them to 0:
+    ```text {linenos=table}
+    kubectl scale deployment mendix-agent --replicas=0
+    k scale deployment mendix-operator --replicas=0
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: change this to :
kubernetes scale deployment mendix-operator --replicas=0

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+2. Create the backup by entering the following command:
+    ```text {linenos=false}
+    velero create backup mendix-velero-bkp
+    ``` 
+    {{% alert color="info" %}}The above command creates a backup of all your namespaces with the name *mendix-velero-bkp*. If you only want to back up a specific namespace, use the `include-namespace` flag.{{% /alert %}}
+3. Verify that the backup is complete by entering the following command:
+    ```text {linenos=false}
+    velero backup describe mendix-velero-bkp
+    ```
+4. Restart the Mendix Operator and Mendix agent.
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Provide the command on how to restart:
kubectl scale deployment mendix-agent --replicas=1
kubectl scale deployment mendix-operator --replicas=1

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+## 4 Restoring a Velero Backup
+
+To restore a backup that you created with Velero, follow these steps:
+
+1. Restore the backup by entering the following command:
+    ```text {linenos=false}
+    velero restore create --from-backup {<backup name>}
+    ```
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md:68]: Also add a text that for restoring a namespace, customer need to use include-namespace` flag.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    {{% alert color="info" %}}If you only want to restore a specific namespace, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: This should be changed to:

{{% alert color="info" %}}If you only want to restore a specific kubernetes resource, say ,mendix App storage instances, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    ``` 
+    {{% alert color="info" %}}The above command creates a backup of all your namespaces with the name *mendix-velero-bkp*. If you only want to back up a specific namespace, use the `include-namespace` flag.{{% /alert %}}
+3. Verify that the backup is complete by entering the following command:
+    ```text {linenos=false}
+    velero backup describe mendix-velero-bkp
+    ```
+4. Restart the Mendix Operator and Mendix agent.
+
+## 4 Restoring a Velero Backup
+
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md:55]: Add section, where we mention:

We are restoring below mendix objects:

storageinstances.privatecloud.mendix.com

storageplans.privatecloud.mendix.com

builds.privatecloud.mendix.com

mendixapps.privatecloud.mendix.com

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+To restore a backup that you created with Velero, follow these steps:
+
+1. Restore the backup by entering the following command:
+    ```text {linenos=false}
+    velero restore create --from-backup {<backup name>}
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Change this command to:

velero restore create --from-backup mendix-velero-bkp --status-include-resources=storageinstances.privatecloud.mendix.com,storageplans.privatecloud.mendix.com,builds.privatecloud.mendix.com,mendixapps.privatecloud.mendix.com

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 1 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: I guess I'm wondering if that's something that we should actively recommend, then? Wouldn't it always make sense to back up your Kubernetes resources?

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    ```
+    {{% alert color="info" %}}If you only want to restore a specific namespace, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}
+2. After the app has started and created the database, [restore a backup](/developerportal/operate/restore-backup/) of your database and S3 files.
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Also add the text in end:

After completing the restore, add finalizers to StorageInstances (this step is optional, but highly recommended to ensure that Kubernetes garbage collection will clean up storage from deleted environments):

`kubectl patch storageinstances $(kubectl get storageinstances --no-headers -o custom-columns=":metadata.name") -p '{"metadata":{"finalizers":["finalizer.privatecloud.mendix.com"]}}' --type=merge`

**Review by Nidhi251289 — COMMENTED** [review]
Kindly check my comments

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 1 above)
Comment (Nidhi251289) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: we can provide a link for this document to our customers
The decision should be upto the customers. Not us. 
This is quite critical, if they make some mistakes and follow our recommendation then it might be an issue

Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Makes sense - perhaps then I'll change the first sentence to "[Velero](https://velero.io/docs/) is a tool that you can use to back up your Kubernetes namespaces as an additional disaster recovery measure." That makes it a little clearer without making it look required. What do you think?

Comment (Nidhi251289) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: yes

Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 2 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Changed "kubectl scale deployment mendix-agent --replicas=0
    k scale deployment mendix-operator --replicas=0"
to kubectl scale deployment mendix-agent --replicas=0
    kubernetes scale deployment mendix-operator --replicas=0

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 3 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 6 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md:55]: I added the following text to the Introduction section (line 14):

"Velero enables you to back up and restore the following Mendix objects:

* `storageinstances.privatecloud.mendix.com`
* `storageplans.privatecloud.mendix.com`
* `builds.privatecloud.mendix.com`
* `mendixapps.privatecloud.mendix.com`"

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 7 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 4 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md:68]: Covered below.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 5 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 8 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    {{% alert color="info" %}}The process of creating a recovery cluster may vary depending on the platform that you use to host your private cloud. For more information, refer to the documentation supplied by your cloud provider.{{% /alert %}}
+
+## 3 Creating a Velero Backup
+
+To create a backup with Velero, follow these steps:
+
+1. Stop the Mendix Operator and Mendix agent by scaling them to 0:
+    ```text {linenos=table}
+    kubectl scale deployment mendix-agent --replicas=0
+    kubernetes scale deployment mendix-operator --replicas=0
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Its not kubernetes its kubectl.
Should be:

`kubectl scale deployment mendix-operator --replicas=0`

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+## 4 Restoring a Velero Backup
+
+To restore a backup that you created with Velero, follow these steps:
+
+1. Restore the backup by entering the following command:
+    ```text {linenos=false}
+    velero restore create --from-backup mendix-velero-bkp --status-include-resources=storageinstances.privatecloud.mendix.com,storageplans.privatecloud.mendix.com,builds.privatecloud.mendix.com,mendixapps.privatecloud.mendix.com
+    ```
+    {{% alert color="info" %}}If you only want to restore a specific Kubernetes resource, such as the Mendix app storage instances, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}
+2. After the app has started and created the database, [restore a backup](/developerportal/operate/restore-backup/) of your database and S3 files.
```
Comment (Nidhi251289) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Change the restore a backup link to:

/developerportal/deploy/private-cloud-data-transfer/

Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md (hunk 9 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Done.

File: content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+## 4 Restoring a Velero Backup
+
+To restore a backup that you created with Velero, follow these steps:
+
+1. Restore the backup by entering the following command:
+    ```text {linenos=false}
+    velero restore create --from-backup mendix-velero-bkp --status-include-resources=storageinstances.privatecloud.mendix.com,storageplans.privatecloud.mendix.com,builds.privatecloud.mendix.com,mendixapps.privatecloud.mendix.com
+    ```
+    {{% alert color="info" %}}If you only want to restore a specific Kubernetes resource, such as the Mendix app storage instances, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}
```
Comment (zlogic) [inline content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: I think it's best to restore statuses for all resources listed above - if a user decides to only restore the status for some resources, this might result in unpredictable behaviour or errors.

Could you rephrase this that *restoring all listed resources is highly recommended*, and that restoring only specific resources could result in unpredictable behaviour?

Comment (katarzyna-koltun-mx) [reply content/en/docs/developerportal/deploy/private-cloud/private-cloud-velero.md]: Changed to:

{{% alert color="warning" %}}As a best practice, it is recommended to restore all resources, as in the above example. Restoring only specific resources can result in unpredictable behavior. However, if you only want to restore a specific Kubernetes resource, use the `--status-include-resources flag`, for example, `--status-include-resources=storageinstances.privatecloud.mendix.com`.{{% /alert %}}
